types/src/v0/impls/stake_table.rs

```diff
@@ -239,168 +215,220 @@ pub fn validators_from_l1_events<I: Iterator<Item = StakeTableEvent>>(
-                };
+                // The contract does *not* enforce that each schnorr key is only used once.
+                if !self.used_schnorr_keys.insert(state_ver_key.clone()) {
+                    tracing::warn!(
+                        schnorr_key = %state_ver_key,
+                        "Duplicate Schnorr key found during stake table events processing"
+                    );
+                    return Err(StakeTableError::SchnorrKeyAlreadyUsed(
+                        state_ver_key.to_string(),
+                    ));
+                }
 
-                bls_keys.insert(stake_table_key);
-                schnorr_keys.insert(state_ver_key.clone());
+                entry.or_insert(Validator {
+                    account,
+                    stake_table_key,
+                    state_ver_key,
+                    stake: U256::ZERO,
+                    commission,
+                    delegators: HashMap::new(),
+                });
 
-                match validators.entry(account) {
-                    indexmap::map::Entry::Occupied(_occupied_entry) => {
-                        bail!("validator {:#x} already registered", *account)
-                    },
-                    indexmap::map::Entry::Vacant(vacant_entry) => vacant_entry.insert(Validator {
-                        account,
-                        stake_table_key,
-                        state_ver_key,
-                        stake: U256::from(0_u64),
-                        commission,
-                        delegators: HashMap::default(),
-                    }),
-                };
+                Ok(())
             },
-            StakeTableEvent::RegisterV2(event) => {
+
+            StakeTableEvent::RegisterV2(reg) => {
                 // Signature authentication is performed right after fetching, if we get an
                 // unauthenticated event here, something went wrong, we abort early.
-                event
-                    .authenticate()
-                    .context("Failed to authenticate event: {event:?}")?;
+                reg.authenticate()
+                    .map_err(|e| StakeTableError::AuthenticationFailed(e.to_string()))?;
+
                 let ValidatorRegisteredV2 {
                     account,
                     blsVK,
                     schnorrVK,
                     commission,
                     ..
-                } = event;
+                } = reg;
 
                 let stake_table_key: BLSPubKey = blsVK.into();
                 let state_ver_key: SchnorrPubKey = schnorrVK.into();
+
                 // The stake table contract enforces that each bls key is only used once.
-                if bls_keys.contains(&stake_table_key) {
-                    bail!("bls key already used: {}", stake_table_key.to_string());
-                };
+                if !self.used_bls_keys.insert(stake_table_key) {
+                    tracing::warn!(
```

Comment:
```suggestion
                    tracing::error!(
```
This should really never happen, so I think we should at least log an error.

Reply:
I have removed this log. we are applying events in. validators_from_l1_events() and there I have added error level log for unexpected error and warn  level for expected error 
https://github.com/EspressoSystems/espresso-network/pull/3396/commits/ba027c88b5fcc7226a6afaa0557b0e401058f097